File: bush-maths/python/test/utils/map_hook_state.py
from src.hooks.stable_surge.types import (
    StableSurgeHookState,
    map_stable_surge_hook_state,
)
from src.hooks.types import HookState


def map_hook_state(hook_data: dict, pool_data: dict) -> HookState:
    """
    Maps hook data from JSON to typed HookState objects.

    This function serves as a central dispatcher that converts raw hook data
    from test JSON files into strongly-typed HookState objects based on the
    hook type.

    Args:
        hook_data: Raw hook dict from JSON with fields:
            - type: Hook type identifier (e.g., "STABLE_SURGE")
            - dynamicData: Hook-specific parameters (varies by type)
            - address: Hook contract address
            - Configuration flags (shouldCallAfterSwap, etc.)
        pool_data: Pool dict with fields needed by hooks:
            - tokens: List of token addresses
            - amp: Amplification parameter (for stable pools with STABLE_SURGE)

    Returns:
        Typed HookState object (StableSurgeHookState)

    Raises:
        ValueError: If hook type is unsupported or required fields are missing
    """
    hook_type = hook_data.get("type")

    if hook_type == "STABLE_SURGE":
        return _map_stable_surge_hook(hook_data, pool_data)
    else:
        raise ValueError(f"Unsupported hook type: {hook_type}")


def _map_stable_surge_hook(hook_data: dict, pool_data: dict) -> StableSurgeHookState:
    """Maps STABLE_SURGE hook data to StableSurgeHookState."""
    if "dynamicData" not in hook_data:
        raise ValueError("STABLE_SURGE hook requires dynamicData")

    dynamic_data = hook_data["dynamicData"]

    if "surgeThresholdPercentage" not in dynamic_data:
        raise ValueError(
            "STABLE_SURGE hook requires surgeThresholdPercentage in dynamicData"
        )
    if "maxSurgeFeePercentage" not in dynamic_data:
        raise ValueError(
            "STABLE_SURGE hook requires maxSurgeFeePercentage in dynamicData"
        )
    if "amp" not in pool_data:
        raise ValueError("STABLE_SURGE hook requires amp from pool data")

    return map_stable_surge_hook_state(
        {
            "surgeThresholdPercentage": dynamic_data["surgeThresholdPercentage"],
            "maxSurgeFeePercentage": dynamic_data["maxSurgeFeePercentage"],
            "amp": pool_data["amp"],
        }
    )

File: bush-maths/rust/tests/utils/map_hook_state.rs
//! Maps hook data from JSON test files into typed HookState objects

use alloy_primitives::U256;
use bush_maths::hooks::types::HookState;
use bush_maths::hooks::StableSurgeHookState;

use super::read_test_data::HookData;

/// Pool data needed for hook state mapping
#[allow(dead_code)]
pub struct PoolData {
    pub tokens: Vec<String>,
    pub amp: Option<U256>,
    pub weights: Option<Vec<U256>>,
    pub swap_fee: U256,
}

/// Maps hook data from JSON test files into typed HookState objects
/// that can be used with hook implementations
pub fn map_hook_state(
    hook_data: &HookData,
    pool_data: &PoolData,
) -> Result<HookState, Box<dyn std::error::Error>> {
    match hook_data.hook_type.as_str() {
        "STABLE_SURGE" => map_stable_surge_hook_state(hook_data, pool_data),
        _ => Err(format!("Unsupported hook type: {}", hook_data.hook_type).into()),
    }
}


fn map_stable_surge_hook_state(
    hook_data: &HookData,
    pool_data: &PoolData,
) -> Result<HookState, Box<dyn std::error::Error>> {
    let dynamic_data = &hook_data.dynamic_data;

    let surge_threshold_percentage = dynamic_data["surgeThresholdPercentage"]
        .as_str()
        .ok_or("STABLE_SURGE hook requires surgeThresholdPercentage in dynamicData")?
        .parse::<U256>()?;

    let max_surge_fee_percentage = dynamic_data["maxSurgeFeePercentage"]
        .as_str()
        .ok_or("STABLE_SURGE hook requires maxSurgeFeePercentage in dynamicData")?
        .parse::<U256>()?;

    let amp = pool_data
        .amp
        .ok_or("STABLE_SURGE hook requires amp from pool data")?;

    Ok(HookState::StableSurge(StableSurgeHookState {
        hook_type: "StableSurge".to_string(),
        amp,
        surge_threshold_percentage,
        max_surge_fee_percentage,
    }))
}

